File: travel_agent/app/schemas/trip.py
from __future__ import annotations

from typing import Any

from pydantic import Field

from travel_agent.app.schemas.approvals import ApprovalRequest, BookingRecord
from travel_agent.app.schemas.brief import TripBrief
from travel_agent.app.schemas.budget import BudgetEstimate
from travel_agent.app.schemas.common import (
    AuditEvent,
    CriticFinding,
    SourceRef,
    StrictBaseModel,
    TripStatus,
)
from travel_agent.app.schemas.itinerary import Itinerary
from travel_agent.app.schemas.providers import (
    AccommodationOption,
    FlightOption,
    FxInfo,
    LocalEventsGuide,
    LocalTransportPlan,
    MultiCityPlan,
    NearbyGuide,
    POIOption,
    PrepChecklist,
    SafetyInfo,
    StayAreaGuide,
    TransportTicketGuide,
    VisaCheckResult,
)


class TripPlanState(StrictBaseModel):
    trip_id: str
    user_id: str | None = None
    locale: str = "ko-KR"
    currency: str = "KRW"
    timezone: str = "Asia/Seoul"
    raw_user_message: str
    raw_user_messages: list[str] = Field(default_factory=list)
    brief: TripBrief | None = None
    user_profile_snapshot: dict[str, Any] = Field(default_factory=dict)
    constraints: dict[str, Any] = Field(default_factory=dict)
    missing_fields: list[str] = Field(default_factory=list)
    assumptions: list[str] = Field(default_factory=list)
    destination_candidates: list[str] = Field(default_factory=list)
    selected_destination: str | None = None
    transport_options: list[FlightOption] = Field(default_factory=list)
    accommodation_options: list[AccommodationOption] = Field(default_factory=list)
    poi_candidates: list[POIOption] = Field(default_factory=list)
    activity_options: list[POIOption] = Field(default_factory=list)
    local_transport_options: list[dict[str, Any]] = Field(default_factory=list)
    route_evidence_refs: list[str] = Field(default_factory=list)
    draft_itinerary: Itinerary | None = None
    optimized_itinerary: Itinerary | None = None
    budget: BudgetEstimate | None = None
    risk_findings: list[CriticFinding] = Field(default_factory=list)
    visa_result: VisaCheckResult | None = None
    local_transport: LocalTransportPlan | None = None
    fx_info: FxInfo | None = None
    safety_info: SafetyInfo | None = None
    nearby_guide: NearbyGuide | None = None
    stay_area_guide: StayAreaGuide | None = None
    prep_checklist: PrepChecklist | None = None
    multicity_plan: MultiCityPlan | None = None
    local_events: LocalEventsGuide | None = None
    transport_tickets: TransportTicketGuide | None = None
    critic_findings: list[CriticFinding] = Field(default_factory=list)
    approval_requests: list[ApprovalRequest] = Field(default_factory=list)
    booking_records: list[BookingRecord] = Field(default_factory=list)
    source_refs: list[SourceRef] = Field(default_factory=list)
    evidence_refs: list[str] = Field(default_factory=list)
    audit_log: list[AuditEvent] = Field(default_factory=list)
    # 대화형 질문("오타루 볼거 뭐있냐")에 LLM이 바로 답한 텍스트(계획 요약 대신 표시).
    assistant_message: str | None = None
    # 정보가 부족할 때 '무엇을 알려주면 더 정확해지는지' 제안 문장(LLM 우선, 규칙 기반 폴백).
    clarification: str | None = None
    status: TripStatus = TripStatus.intake

    @property
    def primary_destination(self) -> str | None:
        """일정·횡단정보의 기준 도시. 선택된 목적지 우선, 없으면 brief 첫 후보.

        13곳에 흩어져 있던 `selected_destination or (brief.destinations[0] ...)` 보일러플레이트를
        한 곳으로 모아, 멀티시티/도시전환 규칙이 바뀌어도 여기만 고치면 되게 한다.
        property라 Pydantic 필드가 아니어서 직렬화(응답 shape)에 영향 없다.
        """
        if self.selected_destination:
            return self.selected_destination
        if self.brief and self.brief.destinations:
            return self.brief.destinations[0]
        return None


class TripCreateRequest(StrictBaseModel):
    message: str
    user_id: str | None = None
    locale: str = "ko-KR"
    currency: str = "KRW"
    timezone: str = "Asia/Seoul"


class TripMessageRequest(StrictBaseModel):
    message: str


class TripSummaryResponse(StrictBaseModel):
    trip_id: str
    status: TripStatus
    summary: str
    missing_fields: list[str] = Field(default_factory=list)
    questions: list[str] = Field(default_factory=list)
    state: TripPlanState | None = None


class FinalPlanResponse(StrictBaseModel):
    trip_id: str
    status: TripStatus
    summary: str
    assumptions: list[str]
    missing_fields: list[str]
    recommended_destination: str | None
    transport_options: list[FlightOption]
    accommodation_options: list[AccommodationOption]
    itinerary: Itinerary | None
    budget: BudgetEstimate | None
    visa_result: VisaCheckResult | None = None
    local_transport: LocalTransportPlan | None = None
    fx_info: FxInfo | None = None
    safety_info: SafetyInfo | None = None
    nearby_guide: NearbyGuide | None = None
    stay_area_guide: StayAreaGuide | None = None
    prep_checklist: PrepChecklist | None = None
    multicity_plan: MultiCityPlan | None = None
    local_events: LocalEventsGuide | None = None
    transport_tickets: TransportTicketGuide | None = None
    risk_findings: list[CriticFinding]
    critic_findings: list[CriticFinding]
    approval_requests: list[ApprovalRequest]
    source_refs: list[SourceRef]
    next_actions: list[str]

File: frontend/src/types/trip.ts
import type { ApprovalRequest, BookingRecord } from './approval'
import type { BudgetEstimate } from './budget'
import type { CriticFinding, Location, Money, ProviderMetadata, SourceRef, TripStatus } from './common'
import type { Itinerary } from './itinerary'

export interface TripBrief {
  origin?: string | null
  destinations: string[]
  start_date?: string | null
  end_date?: string | null
  flexible_dates: boolean
  duration_days?: number | null
  travelers?: number | null
  budget_total?: number | null
  budget_per_person?: number | null
  currency: string
  travel_style?: string | null
  pace?: string | null
  accommodation_preference?: string | null
  transport_preference?: string | null
  accessibility_needs: string[]
  dietary_restrictions: string[]
  passport_country?: string | null
  visa_status_known: boolean
  must_include: string[]
  must_avoid: string[]
  missing_fields: string[]
  assumptions: string[]
}

export interface FlightOption {
  option_id: string
  airline: string
  origin: string
  destination: string
  departure_time: string
  arrival_time: string
  return_departure_time?: string | null
  return_arrival_time?: string | null
  price: Money
  refundable: boolean
  booking_required: boolean
  metadata: ProviderMetadata
  notes: string[]
}

export interface AccommodationOption {
  option_id: string
  name: string
  location: Location
  nightly_price: Money
  total_price: Money
  rating?: number | null
  star_rating?: number | null
  review_count?: number | null
  amenities?: string[]
  cancellation_policy: string
  metadata: ProviderMetadata
  notes: string[]
}

export interface POIOption {
  poi_id: string
  title: string
  type: string
  location: Location
  area: string
  estimated_cost: Money
  rating?: number | null
  review_count?: number | null
  opening_hours?: string | null
  recommended_duration_minutes: number
  booking_required: boolean
  metadata: ProviderMetadata
  notes: string[]
}

export interface TripPlanState {
  trip_id: string
  user_id?: string | null
  locale: string
  currency: string
  timezone: string
  raw_user_message: string
  raw_user_messages: string[]
  brief?: TripBrief | null
  user_profile_snapshot: Record<string, unknown>
  constraints: Record<string, unknown>
  missing_fields: string[]
  assumptions: string[]
  destination_candidates: string[]
  selected_destination?: string | null
  transport_options: FlightOption[]
  accommodation_options: AccommodationOption[]
  poi_candidates: POIOption[]
  activity_options: POIOption[]
  local_transport_options: Array<Record<string, unknown>>
  route_evidence_refs: string[]
  draft_itinerary?: Itinerary | null
  optimized_itinerary?: Itinerary | null
  budget?: BudgetEstimate | null
  visa_result?: VisaCheckResult | null
  local_transport?: LocalTransportPlan | null
  fx_info?: FxInfo | null
  safety_info?: SafetyInfo | null
  nearby_guide?: NearbyGuide | null
  stay_area_guide?: StayAreaGuide | null
  prep_checklist?: PrepChecklist | null
  transport_tickets?: TransportTicketGuide | null
  risk_findings: CriticFinding[]
  critic_findings: CriticFinding[]
  approval_requests: ApprovalRequest[]
  booking_records: BookingRecord[]
  source_refs: SourceRef[]
  evidence_refs: string[]
  assistant_message?: string | null
  clarification?: string | null
  status: TripStatus
}

export interface LocalTransportItem {
  category: string
  name: string
  detail?: string | null
  price?: string | null
  duration?: string | null
  source_url?: string | null
}

export interface LocalTransportPlan {
  city: string
  summary: string
  airport_transfers: LocalTransportItem[]
  transit_passes: LocalTransportItem[]
  tips: string[]
  source_url?: string | null
}

export interface BookingPlatform {
  name: string
  url: string
  covers: string
  note?: string | null
}

export interface RouteLink {
  label: string
  maps_url: string
  booking_url?: string | null
}

export interface PassSuggestion {
  name: string
  url: string
  note: string
}

export interface TransportTicketGuide {
  destination_country: string
  summary: string
  hub?: string | null
  hub_lat?: number | null
  hub_lng?: number | null
  platforms: BookingPlatform[]
  pass_suggestion?: PassSuggestion | null
  route_links: RouteLink[]
  source_note: string
}

export interface NearbyDestination {
  name: string
  travel_time: string
  transport: string
  highlights: string[]
  best_for?: string | null
  source_url?: string | null
}

export interface NearbyGuide {
  hub: string
  summary: string
  destinations: NearbyDestination[]
  source_url?: string | null
}

export interface StayArea {
  name: string
  vibe: string
  good_for: string[]
  note?: string | null
  source_url?: string | null
}

export interface StayAreaGuide {
  destination: string
  summary: string
  areas: StayArea[]
  source_url?: string | null
}

export interface PrepGroup {
  title: string
  items: string[]
}

export interface PrepChecklist {
  destination: string
  summary: string
  groups: PrepGroup[]
}

export interface EmergencyContact {
  label: string
  number: string
}

export interface SafetyInfo {
  destination_country: string
  summary: string
  emergency_contacts: EmergencyContact[]
  consular_call_center: string
  embassy_note?: string | null
  travel_advisory?: string | null
  insurance_tips: string[]
  local_cautions: string[]
  source_url?: string | null
}

export interface FxSample {
  local_label: string
  krw_label: string
}

export interface FxInfo {
  base_currency: string
  target_currency: string
  target_per_base: number
  base_per_target: number
  samples: FxSample[]
  budget_total_base?: number | null
  budget_total_target?: number | null
  budget_total_target_label?: string | null
  tips: string[]
  source_url?: string | null
}

export interface VisaCheckResult {
  destination_country: string
  summary: string
  requires_official_verification: boolean
  missing_required_info: string[]
  passport_country?: string | null
  visa_required?: boolean | null
  visa_free_days?: number | null
  entry_authorization?: string | null
  passport_validity_rule?: string | null
  details: string[]
  source_url?: string | null
}

export interface TripCreateRequest {
  message: string
  user_id?: string | null
  locale: string
  currency: string
  timezone: string
}

export interface TripMessageRequest {
  message: string
}

export interface TripSummaryResponse {
  trip_id: string
  status: TripStatus
  summary: string
  missing_fields: string[]
  questions: string[]
  state?: TripPlanState | null
}

export interface FinalPlanResponse {
  trip_id: string
  status: TripStatus
  summary: string
  assumptions: string[]
  missing_fields: string[]
  recommended_destination?: string | null
  transport_options: FlightOption[]
  accommodation_options: AccommodationOption[]
  itinerary?: Itinerary | null
  budget?: BudgetEstimate | null
  risk_findings: CriticFinding[]
  critic_findings: CriticFinding[]
  approval_requests: ApprovalRequest[]
  source_refs: SourceRef[]
  next_actions: string[]
}
